File: DesignPatternDemo/Pattern/Visitor/Employee.m
//
//  Employee.m
//  DesignPatternDemo
//
//  Created by ruantong on 2019/5/11.
//  Copyright © 2019 wupeng. All rights reserved.
//

#import "Employee.h"

@implementation Employee

- (void)report{
    NSString * info = [NSString stringWithFormat: @"姓名:%@",_name];
    info = [NSString stringWithFormat:@"%@ 性别:%@",info,_sex == 1?@"女":@"男"];
    info = [NSString stringWithFormat:@"%@ 薪水:%ld",info,_salary];
    info = [NSString stringWithFormat:@"%@ %@",info,[self getOtherInfo]];
    NSLog(@"%@",info);
}

@end

File: DesignPatternDemo/Pattern/Visitor/Employee.swift
//
//  Employee.swift
//  DesignPatternDemo
//
//  Created by wupeng on 2020/3/4.
//  Copyright © 2020 wupeng. All rights reserved.
//

import Foundation

class Employee: NSObject {
    var name:NSString
    var salary:Double
    var sex:NSString
    
    init(name:NSString,salary:Double,sex:NSString) {
        self.name = name
        self.salary = salary
        self.sex = sex
    }
    
    func accept(visitor:IVisitor) {
        
    }
}

class CommonEmployee: Employee {
    var job:NSString = ""
    
    init(name: NSString, salary: Double, sex: NSString,job:NSString) {
        super.init(name: name, salary: salary, sex: sex)
        self.job = job
    }
    
    override func accept(visitor: IVisitor) {
        visitor.visitCommonEmployee(commonEmployee: self)
    }
}

class Manager: Employee {
    var performance:NSString = "" //业绩
    
    init(name: NSString, salary: Double, sex: NSString,performance:NSString) {
        super.init(name: name, salary: salary, sex: sex)
        self.performance = performance
    }
    
    override func accept(visitor: IVisitor) {
        visitor.visitManager(manager: self)
    }
}
